File: C++/0213-house-robber-ii.cpp
class Solution {
  int rob_simple(vector<int>& nums, int start, int end) {
    int t1 = 0;
    int t2 = 0;
    for (int i = start; i <= end; i++) {
      int current = nums[i];
      int temp = t1;
      t1 = max(current + t2, t1);
      t2 = temp;
    }
    return t1;
  }

 public:
  int rob(vector<int>& nums) {
    int total = nums.size();
    if (total == 0) return 0;
    if (total == 1) return nums[0];
    if (total == 2) return max(nums[0], nums[1]);
    int max1 = rob_simple(nums, 0, nums.size() - 2);
    int max2 = rob_simple(nums, 1, nums.size() - 1);
    return max(max1, max2);
  }
};

File: Python/0213-house-robber-ii.py
# time complexity: O(n)
# space complexity: O(n)
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        def robMoney(money: List[int]):
            dp = [0 for _ in range(len(money) + 1)]
            dp[0] = 0
            dp[1] = money[0]
            for i in range(2, len(money) + 1):
                dp[i] = max(dp[i-2] + money[i-1], dp[i-1])
            return dp[-1]

        n = len(nums)
        if n == 0:
            return 0
        if n == 1:
            return nums[0]
        return max(robMoney(nums[:-1]), robMoney(nums[1:]))


nums = [2, 3, 2]
print(Solution().rob(nums))
nums = [1, 2, 3, 1]
print(Solution().rob(nums))
nums = [1, 2, 3]
print(Solution().rob(nums))
